src/lib.rs

```diff
@@ -352,6 +352,86 @@ impl Filter {
     }
 }
 
+/// The possible input methods.
+pub enum InputMethod {
+    /// Read the input from the local Git history.
+    GitHistory,
+
+    /// Read the specified input file.
+    LogFile(String),
+
+    /// There was no input method specified.
+    None,
+
+    /// Read from `stdin`.
+    Stdin,
+}
+
+impl InputMethod {
+    /// Create a new instance from the given command line options.
+    pub fn new(matches: &getopts::Matches) -> Self {
+        if matches.opt_present("git") {
+            Self::GitHistory
+        } else if matches.opt_present("input") {
+            Self::LogFile(matches.opt_str("input").unwrap())
+        } else if matches.opt_present("stdin") {
+            Self::Stdin
+        } else {
+            Self::None
+        }
+    }
+
+    /// Process the configured input method.
+    pub fn read(&self) -> Option<String> {
```

Comment:
I think it would be nice to let `read` return a `Result<String, Box<dyn Error>>`, then you could also use the `?` in this function.

Also a note for the future, instead of
```rust
match expression {
    Ok(it) => it,
    Err(_) => return None,
}
```
you can also just write
```
expression.ok()?
```

Reply:
I will also incorporate the fixes suggested in #24 in order to fix the named typos. :wrench:

Reply:
The use of `?` would require the definition of a specialised error enum since Rust cannot deduce an appropriate error type from the return type.  The suggest change would just affect types to match against.

Reply:
The change would not be meaningful.  It would allow a more detailed error handling, for sure, but at the moment, the current error handling would just discard the error types and print the already implemented messages.  Hence, I am not going to change this.

Reply:
> The use of `?` would require the definition of a specialised error enum since Rust cannot deduce an appropriate error type from the return type. The suggest change would just affect types to match against.

What? Why would it not be able to deduce? The Err cases are all either a `std::io::Error` which implements the `Error` trait or a `std::string::FromUtf8Error` which also implements the `Error`trait. I do not see, where this would fail to compile. 😅 

> The change would not be meaningful. It would allow a more detailed error handling, for sure, but at the moment, the current error handling would just discard the error types and print the already implemented messages. Hence, I am not going to change this.

It would also make it more clear, that an error is the reason nothing is returned. Otherwise it would be read as something, which just has not always a value.

Reply:
I tried out to implement your suggestion and it did not work.  The thus required workaround added too much overhead that it was not a constructive change anymore.  Hence, I left it as it was.

If you know a working implementation of your suggestion, please post it here so that I can incorporate it.

Reply:
Another option is to make this discussion an issue for a future Pull Request to solve.  That way, we can focus on this aspect later on more in detail.  The suggested implementation implies a correct result, hence, it can be considered working.  And it compiles.  Merging working changes into `main` is in my opinion more important than how it works under the hood.

Reply:
> Otherwise it would be read as something, which just has not always a value.

And yes, the input has not always a value.  So reading it that way is what I wanted to express.  Look up `Maybe` in the Haskell Report.

Reply:
I will update you, when I had time to write one. 😄

Reply:
> > Otherwise it would be read as something, which just has not always a value.
> 
> And yes, the input has not always a value. So reading it that way is what I wanted to express. Look up `Maybe` in the Haskell Report.

But if it has no vaue, it is an error, that is what Result is for. But I will post my suggestion tomorrow

Reply:
:+1:

Reply:
```suggestion
    pub fn read(&self) -> Result<String, Box<dyn std::error::Error>> {
        match self {
            Self::GitHistory => {
                let process = std::process::Command::new("git")
                    .arg("log")
                    .arg("--numstat")
                    .output()?;

                Ok(String::from_utf8(process.stdout)?)
            }
            Self::LogFile(string) => Ok(std::fs::read_to_string(string)?),
            Self::Stdin => {
                let mut input = String::new();

                loop {
                    let mut buffer = String::new();
                    if std::io::stdin().read_line(&mut buffer)? == 0 {
                        break;
                    }
                    input.push_str(&buffer);
                }

                Ok(input)
            }
        }
    }

```

Reply:
Thank you for your suggestion!

I pasted it to the source file and committed it almost unchanged; I just added a single empty line each before and after the `if` statement within the `loop`.  In my opinion, this helps to keep the source file more readable since statements like `if`, `for`, `loop` and so on often have complex bodies which rather should be separated visibly by newlines from adjacent code sections.